### PR TITLE
LS support for other number formatting

### LINKED ISSUE #2127: LS support for other number formatting
**Is your feature request related to a problem? Please describe.**
No

**Describe the solution you'd like**
Current `ls` looks like this:
```
────┬────────────────────┬──────┬──────────┬─────────────
 #  │        name        │ type │   size   │  modified
────┼────────────────────┼──────┼──────────┼─────────────
 10 │ CODE_OF_CONDUCT.md │ File │   3.5 KB │ 6 hours ago
 11 │ CONTRIBUTING.md    │ File │   1.4 KB │ 6 hours ago
 12 │ Cargo.lock         │ File │ 121.5 KB │ 6 hours ago
 13 │ Cargo.toml         │ File │   4.7 KB │ 6 hours ago
 14 │ LICENSE            │ File │   1.1 KB │ 6 hours ago
 15 │ Makefile.toml      │ File │    473 B │ 6 hours ago
 16 │ README.build.txt   │ File │    193 B │ 6 hours ago
 17 │ README.md          │ File │  16.3 KB │ 6 hours ago
```
I'd like to be able to specify how the bytes are displayed whether it's `16.3 KB` or `16279` or `16,279` or even perhaps `15.9 KiB` or `0.015 MiB`. The point is, I'd like more control over how these numbers are represented.

**Describe alternatives you've considered**
`ls | each {to json} | each {from json}`
```────┬────────────────────┬──────┬────────┬───────────────────────────────────
 #  │        name        │ type │  size  │             modified
────┼────────────────────┼──────┼────────┼───────────────────────────────────
 0  │ .azure             │ Dir  │      0 │ 2020-07-06 13:27:12.875672900 UTC
 1  │ .cargo             │ Dir  │      0 │ 2020-07-06 13:27:12.877672400 UTC
 2  │ .circleci          │ Dir  │      0 │ 2020-07-06 13:27:12.879672700 UTC
 3  │ .dockerignore      │ File │      8 │ 2020-07-06 13:27:12.880672900 UTC
 4  │ .editorconfig      │ File │    236 │ 2020-07-06 13:27:12.881673800 UTC
 5  │ .github            │ Dir  │      0 │ 2020-07-06 13:27:12.885672300 UTC
 6  │ .gitignore         │ File │    303 │ 2020-07-06 13:27:12.888672800 UTC
 7  │ .gitpod.Dockerfile │ File │    302 │ 2020-07-06 13:27:12.889672500 UTC
 8  │ .gitpod.yml        │ File │    891 │ 2020-07-06 13:27:12.890673300 UTC
 9  │ .theia             │ Dir  │      0 │ 2020-07-06 13:27:12.892672800 UTC
 10 │ CODE_OF_CONDUCT.md │ File │   3520 │ 2020-07-06 13:27:12.894672900 UTC
 11 │ CONTRIBUTING.md    │ File │   1403 │ 2020-07-06 13:27:12.895670800 UTC
 12 │ Cargo.lock         │ File │ 121461 │ 2020-07-06 13:27:12.897670800 UTC
 13 │ Cargo.toml         │ File │   4721 │ 2020-07-06 13:27:12.898671800 UTC
 14 │ LICENSE            │ File │   1113 │ 2020-07-06 13:27:12.899673 UTC
 15 │ Makefile.toml      │ File │    473 │ 2020-07-06 13:27:12.900683400 UTC
 16 │ README.build.txt   │ File │    193 │ 2020-07-06 13:27:12.901671700 UTC
 17 │ README.md          │ File │  16279 │ 2020-07-06 13:27:12.902672500 UTC
```

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by luccasmmg**
I think it Closes #2127 

Added a flag to the ls command(--size-format/-f) that allows you to define a custom format, currently there are only the file formats, although in the issue related @fdncred talked about other formats that didn't involved a filesize format, so i would like his input on this. 
I have one big problem with my code currently, which is the "convert_bytes_to_string_using_format" function, which i think has the potential to get quite big, in a file(filesystem_shell.rs) which i think already is too big to be easily understandable + i don't know if the filesystem_shell is the best file to put this function, given that it only relates to the ls command. Should i put it in the ls.rs file and just import in the filesystem_shell? Any help would be appreciated.

This is an example of the new flag. The todo file aparently has 0 size so it becomes this "-"
image

### CODE REVIEW

**Comment by fdncred** [comment]
This is very cool. Thanks! I can't wait to play with this. 💯 
 @andrasio or @jonathandturner can you chime in on his question below please?

> I have one big problem with my code currently, which is the "convert_bytes_to_string_using_format" function, which i think has the potential to get quite big, in a file(filesystem_shell.rs) which i think already is too big to be easily understandable + i don't know if the filesystem_shell is the best file to put this function, given that it only relates to the ls command. Should i put it in the ls.rs file and just import in the filesystem_shell? Any help would be appreciated.

File: crates/nu-cli/src/shell/filesystem_shell.rs [hunk 1]
```diff
@@ -1099,3 +1114,44 @@ fn path_contains_hidden_folder(path: &PathBuf, folders: &[PathBuf]) -> bool {
...
+
+fn convert_bytes_to_string_using_format(
+    bytes: u64,
+    format: Tagged<String>,
+) -> Result<UntaggedValue, ShellError> {
+    let byte = byte_unit::Byte::from_bytes(bytes as u128);
+    if byte.get_bytes() == 0u128 {
+        return Ok(UntaggedValue::string("—".to_string()));
+    }
+    match format.item().as_str() {
```
Comment (fdncred) [inline crates/nu-cli/src/shell/filesystem_shell.rs:1126]: One quick comment here. You could make `format.item().as_str()` lower case and then the patterns you try to match could only be "b", "kb", "mb", etc.

Comment (luccasmmg) [reply crates/nu-cli/src/shell/filesystem_shell.rs:1126]: fair enough, will change

**Comment by jonathandturner** [comment]
I was wondering if we might want this to be a more general command, so that other commands (like `df` for example) could also use it. You could run `ls` then pass it through your number format command, which updates the output. In the future that combination could then be aliased back to `ls` and you'd get the end result you want.

**Comment by luccasmmg** [comment]
Im pretty inexperienced  with the project and Rust, so i don't think i could do this at least in my current level, i thought about using this function to add flags into other commands(like `du`). My main problem with making a separate command that takes as output a table, would be that there would be a lot of data lost, no? The Filesize value currently rounds up the bytes to the biggest format possible. So if we tried to convert from GB to B, KB, KiB, MB and MiB. The `from_str()` function in the byte_unit crate would miss some data. Other than that, is there any similar command that takes a table and converts some row that i could take inspiration from?

**Comment by jonathandturner** [comment]
@luccasmmg - there are a couple things that might help here:

The separate command could take as in put the table, like you're saying. It would be able to update data in that table as it comes into the command so that what goes out has already had the value replaced. This is how the `update` command works, for example. You could look at that command. You don't even need the `block` part of update, just the part that updates a value given a path.

If you're talking about converting from the filesize byte value to a string, yes, you'll lose some data. This is also the case for adding a flag. Are you seeing additional places, too? I was thinking this command would be a way to format the table's data for display. We could start it with the filesize functionality you're working on and over time grow from there.

**Comment by luccasmmg** [comment]
I think i'm starting to understand @jonathandturner, could you give me 1 - the name you would like(since format is already taken) 2 - some sort of example of how it would look like? i assume it would be something like this `ls | command_name the_column_to_update the_format` which in this specific case would be `ls | command_name size KB`. I don't really know how much it would take me to build this, but i would like to try in the next weeks if its possible.

**Comment by jonathandturner** [comment]
@luccasmmg you could try it as sub-command of format, maybe something like:

```
ls | format filesize size KiB
```

and since it's focused on just filesize, then you could have the filesize formats you want to support as part of it.  I'd just hardcode them, yeah.

**Comment by luccasmmg** [comment]
Ok, makes sense, should i close this pul request?

**Comment by jonathandturner** [comment]
@luccasmmg if you want or you could take what you have and continue to evolve it. When we land it, we'll squash the commits so we'll only see the final set of changes.

**Comment by fdncred** [comment]
Just throwing my two cents in here. I don't want to have to type 30 characters to get my `ls filesize` in a different format. So, could we have it both ways?

1. An external command so one could do this `ls | format filesize size KiB`
2. A way for `ls` to call that command via code so I can just do `ls -f KiB`

Another solution to this would be for nushell to be able to alias internal commands - then I could have an alias like:
`alias ls [path] { ls $path | format filesize size KiB }`

I guess maybe I could also have an alias like:
`alias KiB [] { format filesize size KiB }`
and then I assume something like this would work
`ls | KiB`

@jonathandturner Thoughts on these items?

**Comment by fdncred** [comment]
@luccasmmg while waiting for @jonathandturner to chime in, I've compiled tried your PR. Here's some notes.

1. When I do `ls -f b` items are listed in bytes but they are left justified and are no longer a `Primitive::Filesize`. I can tell because we can color data types now and they're all the same color as strings. Is there a way to maintain the formatting and have them still be a filesize type? FYI - The reason they are left justified is also because they're a string, if we can change them to filesize they won't be left justified any longer. I remember that in the code.
2. I'd like to see separators in the numbers if possible when doing a `ls -f b`. So, in the US it would be like `1,000,000 B`. Other cultures have different separators. This may not be easily do-able, I'm not sure.
3. For completeness, I'd like to see all the formats that the crate supports. For instance, you have kb & kib, mb & mib, gb but no gib, adn the same for tb, pb, ex, etc, etc.

Generally though, I love this change and i think it's cool that i can do `ls -f b` then `ls -f kb` then `ls -f mb` then `ls -f gb`. Good work!

**Comment by luccasmmg** [comment]
@fdncred So answering to your questions

1 - "When I do ls -f b items appear to be listed in bytes but they are left justified and are no longer a Primitive::Filesize data type" the problem is basically on how the `Primitive:Filesize` is handled, it stores just the number of bytes, the convertion to some format is done way after the command is run. In this piece of code right here `crates/nu-data/src/base/shape.rs` given that, i don't know how i would be able to keep the `Primitive::Filesize` together with some information about how it must be formatted, without basically breaking the whole source code and every instance of `Primitive::Filesize`s if you have any thoughts on this.
image
2 - Do you imply this just for the Bytes format? If so i don't think that would be too difficult if we keep not using the filesize primitive, it would be a simple matter of string formatting.
2.1 -  "we should drop the B. Thoughts?" how would that be? `ls -f` would automatically format into Bytes? If so could you point me to some command which has a flag with similar behaviour(meaning that it outputs an error if the wrong string is typed, while still having a default value) i actually just started using Nu daily quite recently so i don't know many commands, i'm mainly contributing to learn Rust and new things.
3 - I see no problem in adding the other formats.

Currently i'm trying to work on the @jonathandturner idea of a subcommand, i don't have something to show yet firstly because i'm quite busy this week and probably will be until next monday, and secondly because i'm still quite inexperience with the code, i think i found a good north in the `inc` plugin, which it seems to be quite similar to what i'm trying to achieve(map a row instead of just substituting like update does).

I think that wraps up why the problems that you pointed are there(would love to know if you have any solution) and why i couldn't finish the subcommand.

**Comment by fdncred** [comment]
Related to item:
1. This will take @andrasio or @jonathandturner to chime in because I'm not sure.
2. Yes, for bytes only.
2.1 I wasn't clear. I mean to still use `ls -f b` but when it shows up in the listing remove the B from each of the byte sizes since it's already implied.

Thanks for all your help. Keep up the good work!

**Comment by fdncred** [comment]
@luccasmmg we have the `num_format` crate in the `from.rs` file but the comments indicate that there are some problems with it - or perhaps it's different from `num-format`. I'm researching filesize so we can still return a filesize primitive.

**Comment by fdncred** [comment]
The longer I look at this, the more I think that this is what needs to change.
```rust
            InlineShape::Bytesize(bytesize) => {
                let byte = byte_unit::Byte::from_bytes(*bytesize as u128);

                let byte = byte.get_appropriate_unit(false);

                match byte.get_unit() {
                    byte_unit::ByteUnit::B => {
                        (b::primitive(format!("{}", byte.get_value())) + b::space() + b::kind("B"))
                            .group()
                    }
                    _ => b::primitive(byte.format(1)),
                }
            }
```
and the only thing that jumps out at me is moving your formatting function inside this one and then conveying the parameters to this function somehow. Thoughts?

**Comment by luccasmmg** [comment]
I thought about that, and i agree is the only way, but, what we would have to do is add a new value to the `Filesize` primitive with the desired formatting something like `Filesize(u64, Option<ByteUnit>)` where it would automatically decide the formatting if the Option outputs None. BUT, this solution would seriously break stuff, pretty much every use of the FileSize primitive would have to be refactored. that would be a lot of changes, with a lot of potential to bugs. Do you think that is worth the problems?
Also, i added the bytes formatting like you asked, it seems the problem with the other uses was related to platform specific(windows and unix) locale detecting, i decided to just hardcode the `us` standard since its the one most people are at least somewhat familiar. Given that the current choice seems to be between no formatting and just `us` formatting, would you agree? I guess i had to swallow my latin american pride to just use the american format.

**Comment by fdncred** [comment]
@luccasmmg I talked to @jonathandturner about this PR and, when he gets a free moment, he's going to look at it. I think the direction we may go is to land this PR since you've made it a subcommand on format. Good work. In the meantime, I'm going to look at maybe creating a config value that would allow ls to maintain a filesize but formatted in a variety of ways by making that change in the InlineSHape::Bytesize() code we've both mentioned. Hopefully this all sounds ok to you. Not sure I can do that but I'll look into it. Thanks again for all your hard work!

**Comment by luccasmmg** [comment]
No problems on my part, but one question, do you know why the gitpod check is stuck since yesterday? Did that ever happen to you?

**Comment by fdncred** [comment]
re: gitpod. yes, i've seen it a few times. You just need to submit an empty change or some minor change to get the checks to start over again.

**Comment by jonathandturner** [comment]
This looks good, thanks for putting it together!

**Pull request merged**